### PR TITLE
Add brief documentation of overlay classes

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**
It could be expanded in future, but it's better than what was there
before

### CODE REVIEW

**Review by NathanJPhillips — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 25fb9befb2b2875816effbb623f62e592b80a994).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84714306

File: jbmc/src/java_bytecode/java_class_loader.cpp [hunk 1]
```diff
@@ -69,8 +69,16 @@ java_class_loadert::parse_tree_with_overlayst &java_class_loadert::operator()(
 }
 
 /// Check if class is an overlay class by searching for `ID_overlay_class` in
-/// its list of annotations. TODO(nathan) give a short explanation about what
-/// overlay classes are.
+/// its list of annotations.
+///
+/// An overlay class is a class with the annotation
+/// \@OverlayClassImplementation. They serve the following purpose. When the JVM
```
Comment (antlechner) [inline jbmc/src/java_bytecode/java_class_loader.cpp]: There should also be some documentation for the other two annotations (`IgnoredMethodImplementation` and `OverlayMethodImplementation`). Could you add a `TODO` for these here or wherever would be a good place to document them?

**Review by antlechner — APPROVED** [review]

File: jbmc/src/java_bytecode/java_class_loader.cpp (hunk 1 above)
Comment (owen-mc-diffblue) [reply jbmc/src/java_bytecode/java_class_loader.cpp]: I've added documentation for them

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 2]
```diff
@@ -118,11 +118,43 @@ class java_bytecode_convert_classt:public messaget
   // see definition below for more info
   static void add_array_types(symbol_tablet &symbol_table);
 
+  /// Check if a method is an overlay method by searching for
+  /// `ID_overlay_method` in its list of annotations.
+  ///
+  /// An overlay method is a method with the annotation
+  /// \@OverlayMethodImplementation. They should only appear in
```
Comment (antlechner) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:125]: :+1: Thanks for explaining this, I wasn't sure if `@OverlayClassImplementation` automatically turned all methods into overlay methods, of if within an overlay class methods could still be declared to be overlay methods or not.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 3]
```diff
@@ -118,11 +118,43 @@ class java_bytecode_convert_classt:public messaget
...
+  /// [overlay classes](\ref java_class_loader.cpp::is_overlay_class). They
+  /// will be loaded by JBMC instead of the corresponding method in the
```
Comment (antlechner) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: :bee: Could you explain what "the corresponding method" is (i.e. method with the same signature in a class of the same name that was loaded before)?

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 4]
```diff
@@ -118,11 +118,43 @@ class java_bytecode_convert_classt:public messaget
...
+  /// Check if a method is an ignored method by searching for
+  /// `ID_ignored_method` in its list of annotations.
+  ///
+  /// An ignored method is a method with the annotation
+  /// \@IgnoredMethodImplementation. They will be ignored by JBMC. They are
+  /// intended for use in
+  /// [overlay classes](\ref java_class_loader.cpp::is_overlay_class), in
+  /// particular for methods which must exist in the overlay class so that
+  /// it will compile, e.g. default constructors, but which we do not want
+  /// to overlay the corresponding method in the
```
Comment (antlechner) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:150]: :bee:

**Review by antlechner — APPROVED** [review]

**Comment by owen-mc-diffblue** [comment]
I've improved the documentation and also added a test for ignored methods. I've done it in the context of an overlay class, although they do not need to be used in that context, because that is what they were created for.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ae0bb0a203fe807074206869620ff9b8cdc98d41).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84818067

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 5]
```diff
@@ -118,11 +118,44 @@ class java_bytecode_convert_classt:public messaget
...
+  /// underlying class. It is an error if a method in
+  /// an overlay class has the same signature as a method in the underlying
+  /// class and it isn't marked as an
+  /// [overlay method](\ref java_bytecode_convert_classt::is_overlay_method)
+  /// or an ignored method.
+  ///
+  /// \param method: a `methodt` object from a java bytecode parse tree
+  /// \return true if the method is an overlay method, else false
```
Comment (NathanJPhillips) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Whoops

File: jbmc/src/java_bytecode/java_class_loader.cpp [hunk 6]
```diff
@@ -69,9 +69,25 @@ java_class_loadert::parse_tree_with_overlayst &java_class_loadert::operator()(
...
+/// searches the classpath for a particular class, it uses the first match,
+/// and ignores any later matches. JBMC, however, will take account of later
+/// matches as long as they are overlay classes. The
+/// first match is then referred to as the underlying class. The
+/// overlay classes can change the methods of the underlying class in the
+/// following ways: adding a field (by having a  new field), adding a method
```
Comment (NathanJPhillips) [inline jbmc/src/java_bytecode/java_class_loader.cpp]: Double space.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 7]
```diff
@@ -544,11 +549,16 @@ void java_bytecode_convert_classt::convert(
   }
   for(const methodt &method : c.methods)
   {
-    if(is_ignored_method(method))
-      continue;
     const irep_idt method_identifier=
       qualified_classname + "." + id2string(method.name)
-        + ":" + method.descriptor;
+      + ":" + method.descriptor;
```
Comment (NathanJPhillips) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Is this clang? Seems like it should be indented to me.

**Review by NathanJPhillips — CHANGES_REQUESTED** [review]
At least one typo

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp (hunk 5 above)
Comment (owen-mc-diffblue) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Whoops

**Review by NathanJPhillips — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3a2a898c0068ca151982972806b7baffec017b6a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84832959
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 41300064067d3cd61b9543076fad88aed48c1071).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84833676

**Pull request merged**